src/divergence_checker.cc

```diff
@@ -0,0 +1,178 @@
+#include "divergence_checker.h"
+
+namespace celerity::detail::divergence_checker_detail {
+bool divergence_block_chain::check_for_divergence() {
+	add_new_hashes();
+
+	const auto [min_hash_count, max_hash_count] = collect_hash_counts();
+
+	if(min_hash_count == 0) {
+		if(max_hash_count != 0 && m_local_nid == 0) {
+			check_for_deadlock();
+		} else if(max_hash_count == 0) {
+			return true;
+		}
+		return false;
+	}
+
+	const per_node_task_hashes task_hashes = collect_hashes(min_hash_count);
+
+	for(int j = 0; j < min_hash_count; ++j) {
+		const divergence_map check_map = create_divergence_map(task_hashes, j);
+
+		// If there is more than one hash for this task, we have a divergence!
+		if(check_map.size() > 1) { reprot_divergence(check_map, j); }
+	}
+
+	clear(min_hash_count);
+
+	return false;
+}
+
+void divergence_block_chain::reprot_divergence(const divergence_map& check_map, const int task_num) {
+	if(m_local_nid == 0) { log_node_divergences(check_map, task_num + static_cast<int>(m_tasks_checked) + 1); }
+
+	// sleep for local_nid * 100 ms such that we have a no lock synchronized output
+	std::this_thread::sleep_for(std::chrono::milliseconds(m_local_nid * 100));
+
+	log_task_record_once(check_map, task_num);
+
+	m_communicator->barrier();
+
+	throw std::runtime_error("Divergence in task graph detected");
+}
+
+void divergence_block_chain::add_new_hashes() {
+	std::lock_guard<std::mutex> lock(m_task_records_mutex);
+	for(size_t i = m_hashes_added; i < m_task_records.size(); ++i) {
+		std::size_t seed = m_local_hashes.empty() ? m_last_hash : m_local_hashes.back();
+		celerity::detail::utils::hash_combine(seed, std::hash<task_record>{}(m_task_records[i]));
+		m_local_hashes.push_back(seed);
+	}
+	m_last_hash = m_local_hashes.empty() ? m_last_hash : m_local_hashes.back();
+	m_hashes_added = m_task_records.size();
+}
+
+void divergence_block_chain::clear(const int min_progress) {
```

Comment:
Naming: This does not clear the chain (similar to what `vector::clear` would do. Maybe something along the lines of `erase_front`, `prune_leading`, or similar?